### ISSUE TITLE
Some non-US keyboard layouts not working

### ISSUE THREAD

**Issue by nshtg** [atom/atom-keymap#37]
Regarding: atom/atom#2600
@kevinsawicki 

Compare US and (for example)
![800px-kb_us-international svg](https://cloud.githubusercontent.com/assets/5219415/3278897/1a409fec-f3db-11e3-8818-37f183c2184e.png)

German keyboard layout
![800px-kb_germany svg](https://cloud.githubusercontent.com/assets/5219415/3278895/1819be56-f3db-11e3-9fce-4082d5effb20.png)

I think its easy to see why [src/helpers.coffee](https://github.com/atom/atom-keymap/blob/b896a16c2b2d21a0a0b8404527879b50e0adcdaa/src/helpers.coffee#L12-L75) is destined to fail horribly.

One example to demonstrate: I try to Toggle comments using `Ctrl-/`. For that I would have to press `Ctrl-Shift-7` as you can see above. It registers `Ctrl-?`.

I don't know how, but your script somehow catches my `Ctrl-Shift-7` (which would be `shift+55`) and makes it an `?` (`191`) where it should make it an `&`.

__Edit:__ What is that; If I press `#` on my keyboard it gets resolved to `3` in the keybinding resolver, but inserted correctly into the text. Something is very very wrong here! :scream:

**Comment by andersennl** [comment]
Regarding the german keyboard layout: maybe this is obvious for some people, but this is only one possible german keyboard layout. At least the german mac keyboard layout is different than the one on the image. I'm not sure if the e.g. US win/mac keyboards differ as well.

**Comment by ScreenDriver** [comment]
@andersennl that's true. For example I am using the german Neo2 Layout. Most of the usual characters like ``=`` and ``:`` and so on are not working in Atom (Mac OS X) :cry:

**Comment by jannickj** [comment]
I don't know if this is the same bug but I can't unbind the | key. Right now it folds any text tree but I need to use that key for haskell :(

**Comment by nathansobo** [comment]
@jannickj Can you paste the expression you're using to unbind?

**Comment by jannickj** [comment]
@nathansobo thanks, and sorry I didn't update my comment, I found out the correct key to unbind using the atom keybind resolver.

apparently unbinding ctrl-alt-[ did the trick to unbind altgr-|
'atom-workspace atom-text-editor:not([mini])':
  'ctrl-alt-[':'unset!'

**Comment by andischerer** [comment]
Maybe this could help, 
i've released a package that tries to fix all the keyboard-localization stuff: https://github.com/andischerer/atom-keyboard-localization´

There is a keymap translation table generator included for fast generation of foreign keyboard maps which can be used by the package. A german key translation map is already included.
Feel free to check it out ;)

**Comment by jannickj** [comment]
a new update of atom, has made it so unsetting it on my own options does not effect it. I have to manually go into 
...\AppData\Local\atom\app-0.183.0\resources\app\keymaps\win32.json
and delete 

    "ctrl-alt-[": "editor:fold-current-row",
every single time a new update of atom comes out (which is quite often)

this is annoying me to no end, can we please just have fold-current-row removed from the keymap until the issue is resolved? :)

**Comment by 50Wliu** [comment]
> this is annoying me to no end, can we please just have fold-current-row removed from the keymap until the issue is resolved? :)

This makes everyone who *doesn't* have this issue (the majority of Atom users) suffer.  Have you tried out @andischerer's package to see if that helps?

**Comment by nathansobo** [comment]
To my knowledge, nothing we've changed recently should affect the ability to `unset!` a binding. Might be worth investigating and filing an issue.

**Comment by jannickj** [comment]
Hmm.. maybe I am not using it correctly
in my keymap.cson i have tried to set

'atom-workspace atom-text-editor:not([mini])':
  'ctrl-alt-[': 'unset!'

but the folding still happens but it just also adds "|" to where my text-cursor is

**Comment by jannickj** [comment]
@50Wliu oh well was worth a try :p 
anyway i did try the package didn't fix it

**Comment by simonbondo** [comment]
This quickfix worked for me (danish layout):
https://github.com/atom/atom-keymap/issues/35#issuecomment-52508869

**Comment by glsg79** [comment]
On version 0.188.0 using  [andischerer](https://github.com/andischerer) [keyboard-localization](https://github.com/andischerer/atom-keyboard-localization) plugin all seems to work fine

**Comment by bentolor** [comment]
Coming here from atom/tree-view#457 . Atom suggests to use Ctrl-\ to toggle the tree view. Inaccessible because I already need to press AltGr (= Alt + Ctrl) to enter "\" alone. 

My proposal was to switch the primary & secondary key binding (Ctrl-K Ctrl-B) of this action for wider audience.

**Comment by Stanzilla** [comment]
@bentolor have you tried https://github.com/andischerer/atom-keyboard-localization yet?

**Comment by stavpup** [comment]
This is so nasty. I updated from .182 to .208 , to find out that I can not write Greek any more (the characters appear as spaces). Good thing I had kept the old binary. (kubuntu 14.04 64 bit).

**Comment by nathansobo** [comment]
@stavpup That sounds like a different issue. @izuzak have you heard about anything like this?

**Comment by izuzak** [comment]
> That sounds like a different issue. @izuzak have you heard about anything like this?

@nathansobo Hmmm, not sure. Maybe it's related to https://github.com/atom/atom/issues/6214 ? @stavpup Can you check that issue, especially the comments related to changing your font?

**Comment by stavpup** [comment]
I run "atom --safe" and greek characters appear normal. So it must be a package that causes the problem, but which one(s)?

**Comment by izuzak** [comment]
@stavpup You should be able to figure that out by disabling packages one by one.

**Comment by stavpup** [comment]
I got it, its the package https://github.com/braver/fonts

**Comment by enzy** [comment]
No progress on this issue for about a year now...
What about [upvoting the related issue in Chromium](https://code.google.com/p/chromium/issues/detail?id=263724) and promoting this problem further?

**Comment by maschs** [comment]
There are other issues on Android for example that are pretty high priority with thousands of comments and google still does not care since more than a year.

**Comment by Fannon** [comment]
+1

I can't use backspace (\) in atom.io. Which makes it sort of unusable.

**Comment by uchuugaka** [comment]
This should not say "Foreign". That is bad grammar, bad taste, bad style. It should say "International" or "Non-U.S. English"

**Comment by nathansobo** [comment]
@uchuugaka You're absolutely right, and I appreciate you pointing this out. I've renamed the issue in a less USA-centric fashion and hope you'll forgive the original title.

**Comment by gucki** [comment]
Had the same problem with 1.0.2 under ubuntu 14.04. Using  https://github.com/andischerer/atom-keyboard-localization solved it.

**Comment by maduron** [comment]
I've downloaded the keyboard-localization package. Where I'm supposed to place it so I can install it? This is my first day trying Atom.

**Comment by andischerer** [comment]
@maduron [See here](https://github.com/andischerer/atom-keyboard-localization#installation)

**Comment by mnquintana** [comment]
@maduron Also take a look at the [Atom Flight Manual](https://atom.io/docs/latest/) - it's a complete guide for those starting out with Atom. In particular you may want to look at https://atom.io/docs/latest/using-atom-atom-packages

**Comment by maduron** [comment]
Thanx, friends @andischerer , @mnquintana ... I'll do the reading...

**Comment by maduron** [comment]
@andischerer , @mnquintana, Thanks again. My es-la keyboard is working now...

**Comment by prine** [comment]
For all the Swiss people with Mac OS X and the @ problem with alt-g. 
The following keymap.cson addition fixed the problem for me:

```
'atom-workspace':
    'alt-g': 'native!'
```

**Comment by mediafreakch** [comment]
@prine Thanks a lot!!!

**Comment by ezequias** [comment]
For Portuguese Asus (ABNT2) keyboards you could add this two lines to the kymap.cson file and it will solve the problem:

'.platform-win32 atom-text-editor, .platform-linux atom-text-editor':
 'ctrl-alt-q': 'unset!'

**Comment by Amleto** [comment]
I'm experiencing some annoying behaviour I think is related to this issue.

I'm using a uk keyboard with en-gb locale. My # is on the same key as tilde (~). When I type '#', '\' is displayed. What should be displayed as ¬ (shift + key under Esc), atom shows as ~. When I type shift + #, I expect ~, but pipe (|) is shown. When I type '|' (shift + back slash), atom agrees!

**Comment by alexanderadam** [comment]
Currently I have the problem, that F9 key on the same system behaves differently.
On the notebook shortcuts run fine but typing it on an attached keyboard results in a not working F9 key.

Is there any possibility to get the name for the configuration of a certain key?

**Comment by Stanzilla** [comment]
As a heads up, this is now on the 2015 Q4 hotlist https://www.chromium.org/developers/design-documents/input-dev/input-objectives

**Comment by 50Wliu** [comment]
Looks like it's already been delayed:

>Fix international keyboard events
Original owner: dtapuska There are a series of international keyboard events that frustrate users. http://crbug.com/25503 http://crbug.com/42627 http://crbug.com/300475 **Punted: blocked on tricky windows issue**

**Comment by Stanzilla** [comment]
-.-

**Comment by KeitIG** [comment]
Damn, so I guess we'll have to wait ...

**Comment by ghtmtt** [comment]
Don't know if this could help, but with a Debian testing machine with an Italian keyboard the comment is made with ``Ctrl-ù`` instead of ``Ctrl-/``

**Comment by NKjoep** [comment]
That also works with Windows.

ctrl - ù  makes comment the same way of ctrl - / with the Italian keyboard

--
Andrea

**Comment by maxpou** [comment]
For french people, here is a part of my keymap.cson. It will solve this problem.

**[** and **]** character:
```cson
'atom-workspace atom-text-editor:not([mini])':
  'ctrl-alt-[': 'unset!'
  'ctrl-alt-]': 'unset!'
```

**}** character:
```cson
'atom-workspace atom-pane':
  'ctrl-alt-=': 'unset!'
```

**Comment by LMLB** [comment]
[Why Ctrl+Alt shouldn’t be used as a shortcut modifier | The Old New Thing](https://blogs.msdn.microsoft.com/oldnewthing/20040329-00/?p=40003)

**Comment by csholmq** [comment]
@LMLB That resonates with my view at least. ctrl- is fine, alt- is fine, shift- with either ctrl and/or alt is fine. But never ctrl-alt-.

Must emphasize @andischerer and his great plugin https://github.com/andischerer/atom-keyboard-localization which totally does the job right! Sure you loose a few shortcuts, but less pain than having to `unset!` a bunch of keybindings.

**Comment by KeitIG** [comment]
Things are moving =)
https://github.com/atom/electron/issues/2310

**Comment by NKjoep** [comment]
so cool bro! :+1:

**Comment by hellobasti** [comment]
Hey, i have the same problem with the german keyboard layout. Is there any solution for this problem?

Thx

**Comment by ccoenen** [comment]
@hellobasti currently there's just workarounds. Much of that discussion is over in #35, basically look for this package and install it: https://atom.io/packages/keyboard-localization it will probably fix the most pressing problems for a german keyboard.

**Comment by BramVanroy** [comment]
For me the localization package at https://github.com/atom/atom-keymap/issues/37#issuecomment-74121595 fixed it. Thanks @andischerer! I am using the `fr_BE` layout. Note that an alias `nl_BE` would be nice. Quick info: one half of Belgium speaks French, the other Dutch - but we all (mostly) use an azerty (French) keyboard. (Oh, and a couple of folks in the East speak German. I am not sure which layout they'd use!)

**Comment by 50Wliu** [comment]
This should be fixed starting with Atom 1.12.0.

**Comment by Albertc** [comment]
Sorry for my ignorance, I tried to read all lines above.

Can anyone explain what to do to solve the problem than when I type ctrl + / " (for comment a line), atoms receives ctrl + &?

What to put and where

Thanks

**Comment by ccoenen** [comment]
@Albertc 
* short-term solution for most of the keyboard problem is this package: https://atom.io/packages/keyboard-localization for most people it's a "install it and be happy" thing.
* long-term solution: Soon(tm) Atom 1.12 should be released, it features a big change in this general area. It may also solve your problem without any extra package.

**Comment by johan-boule** [comment]
I was about to give a try to Atom, but after seeing this issue, I'll wait for a few years more.

**Comment by 50Wliu** [comment]
> I'll wait for a couple more years.

This issue has already been fixed in Atom 1.12.0 (if you decided to read the issue in its entirety, you would have realized that).

**Comment by Tokimon** [comment]
Personally I am using a danish layout and I also have the issue of wrong mapping.
Maybe I missed something in the 1000000 comments spread out into almost as many issues, but why is it that Atom doesn't look at the actual input given by the keyboard. I can write the right chars well enough, but it is simply the Key Binding Resolver that detects the wrong keys and executes an unintentinal action as a result.

Right now I am having the problem that I can't write the `|` char as the Resolver resolves `ctrl-alt-|` (or `altGr-|`) into `ctrl-alt-[` and that is bound as a shortcut for an action. I will use the `unset!` technique, but to me that is just a hack to the real underlaying problem, that keystrokes are not resolved as sent from the keyboard.

I have stated this problem in https://github.com/atom/atom/issues/6091 as well.

**Comment by 50Wliu** [comment]
@Tokimon As I stated right above you, this issue will be fixed in the upcoming Atom 1.12.0 release.

**Comment by johan-boule** [comment]
@50Wliu I'm sorry about the sarcasm, and i'm glad for you if you eventually have a new system released a few days ago that eventually get the keyboard handling right (I did read that). However, it's now going to be rather hard for me to believe it's gonna be OK. So, yes this issue did harm atom's reputation, because keyboard handling cannot be an afterthought for a text editor! It should have been a basic requirement for releasing anything from day one.

**Comment by ccoenen** [comment]
@johan-boule I have called out the atom team on numerous occasions in many of the open issues. I've called them names, I was angry, I pointed fingers at them for "not caring". You know what? It took a while, it certainly took longer than i'd hoped it would, but now it's probably going to be alright. 

There were major blockers in the way of getting this one right. This is also information buried in the 1000000 comments. It's fine for the vast majority of cases. Yes, there might be bugs, but the earlier those get found, the earlier they will get squashed. As far as I can tell, beta releases are for 1.11 - so the beta channel does not yet have the mentioned changes.

**Comment by alberto** [comment]
Sorry for the spam, but seeing the comments above, I just wanted to leave a huge THANK YOU to the atom team and contributors. You've made an incredible OSS editor and I know a lot of people like me that is grateful for that, even if we are quiet. :heart:

**Comment by hildersantos** [comment]
Sorry to "reopen" this issue, but... I'm on Atom 1.2.5, and the US International keyboard layout is not working for some keys. I'd already tried on a fresh Atom install, and on safe mode, without success. Even the keybinding-resolver is not showing the correct shortcut (i.e. the `^` accent, so essential to vim mode).

It was working very well on 1.2. I made a clean OS install and after reinstalled Atom, this issue just popped out.

![nov-26-2016 13-42-27](https://cloud.githubusercontent.com/assets/2336375/20641786/476e1204-b3de-11e6-95ae-2cb8a6e8f582.gif)

Is there anything I should enable?

Thanks!

**Comment by Ben3eeE** [comment]
@hildersantos Please open a new issue for this bug, thanks.

**Comment by selrond** [comment]
Things are not working for me either.

I'm on __MB Pro with macOS Sierra 10.12.6 (16G29)__
Keyboard: __Slovak QWERTZ__
Atom version: __1.20.0__

When I press `cmd+'` (as recognized in Sublime Text), Atom sees `ctrl+§`. Or `cmd+/` (toggles comment in Sublime Text), Atom sees `cmd+'`.

Weird enough, macOS' native Keyboard viewer shows expected keys when typed, just Atom and VSCode seems to interpret them wrongly. 
I suspect Electron has something to do with it... 

I've already spent about 2 hours trying to resolve the issue, but I'm starting to lose my mind...

__Any suggestions?__

**Comment by ccoenen** [comment]
I think your chances are better opening a new ticket. You can of course link here if anything in this lengthy discussion helps understanding or fixing your issue.

**Comment by johan-boule** [comment]
Does it work on vi or emacs ? That may be an alternative.

**Comment by metemaddar** [comment]
This problem is still present I think.
My Locale is Persian. and the key bindings "**Ctrl+C**"/"**Ctrl+V**"/"**Crtl+/**" Don't work in Persian keyboard. It would be appreciated if it could be solved.
Thanks

**Comment by Ben3eeE** [comment]
@metemaddar Please open a new issue and fill out the entire issue template so we can take a look, thanks 🙇
